device/transport/mqtt/src/mqtt.ts

```diff
@@ -251,8 +251,41 @@ export class Mqtt extends EventEmitter implements DeviceTransport {
           disconnect: (disconnectCallback) => {
             this._fsm.transition('disconnecting', disconnectCallback);
           },
-          sendEvent: (topic, payload, options, sendEventCallback) => {
-            this._mqtt.publish(topic, payload, options, (err, result) => {
+          sendEvent: (message, sendEventCallback) => {
+            /*Codes_SRS_NODE_COMMON_MQTT_BASE_16_008: [The `sendEvent` method shall use a topic formatted using the following convention: `devices/<deviceId>/messages/events/`.]*/
+            let topic = this._getEventTopicFromMessage(message);
+            let systemProperties: { [key: string]: string } = {};
+            /*Codes_SRS_NODE_COMMON_MQTT_BASE_16_011: [The `sendEvent` method shall serialize the `messageId` property of the message as a key-value pair on the topic with the key `$.mid`.]*/
+            if (message.messageId) systemProperties['$.mid'] = message.messageId;
+            /*Codes_SRS_NODE_COMMON_MQTT_BASE_16_012: [The `sendEvent` method shall serialize the `correlationId` property of the message as a key-value pair on the topic with the key `$.cid`.]*/
+            if (message.correlationId) systemProperties['$.cid'] = message.correlationId;
+            /*Codes_SRS_NODE_COMMON_MQTT_BASE_16_013: [The `sendEvent` method shall serialize the `userId` property of the message as a key-value pair on the topic with the key `$.uid`.]*/
+            if (message.userId) systemProperties['$.uid'] = message.userId;
+            /*Codes_SRS_NODE_COMMON_MQTT_BASE_16_014: [The `sendEvent` method shall serialize the `to` property of the message as a key-value pair on the topic with the key `$.to`.]*/
+            if (message.to) systemProperties['$.to'] = message.to;
+            /*Codes_SRS_NODE_COMMON_MQTT_BASE_16_015: [The `sendEvent` method shall serialize the `expiryTimeUtc` property of the message as a key-value pair on the topic with the key `$.exp`.]*/
+            if (message.expiryTimeUtc) {
+              const expiryString = message.expiryTimeUtc instanceof Date ? message.expiryTimeUtc.toISOString() : message.expiryTimeUtc;
+              systemProperties['$.exp'] = (expiryString || undefined);
+            }
+            /*Codes_SRS_NODE_DEVICE_MQTT_16_084: [The `sendEvent` method shall serialize the `contentType` property of the message as a key-value pair on the topic with the key `$.ct`.]*/
+            if (message.contentType) systemProperties['$.ct'] = <string>message.contentType;
+            /*Codes_SRS_NODE_DEVICE_MQTT_16_083: [The `sendEvent` method shall serialize the `contentEncoding` property of the message as a key-value pair on the topic with the key `$.ce`.]*/
+            if (message.contentEncoding) systemProperties['$.ce'] = <string>message.contentEncoding;
+
+            const sysPropString = querystring.stringify(systemProperties);
+            topic += sysPropString;
```

Comment:
I did a refactor here that didn't merge well.  this code (from let  systemProperties: to topic += sysPropString duplicates code that was moved into _getEventTopicFromMessage.  I think this code, as written, will concat sysPropString twice, once inside _getEventTopicFromMessage, and once here.

Reply:
good catch!!

---
In reply to: [197220007](https://github.com/Azure/azure-iot-sdk-node/pull/323#discussion_r197220007) [](ancestors = 197220007)